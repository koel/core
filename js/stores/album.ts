/* eslint camelcase: ["error", {properties: "never"}] */
import Vue from 'vue'
import { union, difference, take, orderBy } from 'lodash'

import stub from '@/stubs/album'
import { artistStore } from '.'
import { http } from '@/services'
import { use } from '@/utils'

export const albumStore = {
  stub,
  cache: {} as { [key: string]: Album },

  state: {
    albums: [stub]
  },

  init (albums: Album[]) {
    // Traverse through the artists array and add their albums into our master album list.
    this.all = albums
    this.all.forEach(album => this.setupAlbum(album))
  },

  setupAlbum (album: Album): void {
    const artist = artistStore.byId(album.artist_id)!
    artist.albums = union(artist.albums, [album])

    Vue.set(album, 'artist', artist)
    Vue.set(album, 'info', null)
    Vue.set(album, 'songs', [])
    Vue.set(album, 'playCount', 0)

    this.cache[album.id] = album
  },

  get all () {
    return this.state.albums
  },

  set all (value) {
    this.state.albums = value
  },

  byId (id: number): Album | undefined {
    return this.cache[id]
  },

  byIds (ids: number[]): Album[] {
    const albums = [] as Album[]
    ids.forEach(id => use(this.byId(id), album => albums.push(album!)))
    return albums
  },

  add (albums: Album | Album[]): void {
    (<Album[]>[]).concat(albums).forEach(album => {
      this.setupAlbum(album)
      album.playCount = album.songs.reduce((count, song) => count + song.playCount, 0)
      this.all.push(album)
    })
  },

  purify (): void {
    this.compact()
  },

  /**
   * Remove empty albums from the store.
   */
  compact (): void {
    const emptyAlbums = this.all.filter(album => album.songs.length === 0)
    if (!emptyAlbums.length) {
      return
    }

    this.all = difference(this.all, emptyAlbums)
    emptyAlbums.forEach(album => delete this.cache[album.id])
  },

  getMostPlayed (n: number = 6): Album[] {
    // Only non-unknown albums with actual play count are applicable.
    const applicable = this.all.filter(album => album.playCount && album.id !== 1)
    return take(orderBy(applicable, 'playCount', 'desc'), n)
  },

  getRecentlyAdded (n: number = 6): Album[] {
    const applicable = this.all.filter(album => album.id !== 1)
    return take(orderBy(applicable, 'created_at', 'desc'), n)
  },

  /**
   * Upload a cover for an album.
   *
   * @param {Album} album The album object
   * @param {string} cover The content data string of the cover
   */
  uploadCover: async (album: Album, cover: string): Promise<string> => {
    const { coverUrl } = await http.put<{ coverUrl: string }>(`album/${album.id}/cover`, { cover })
    album.cover = coverUrl
    return album.cover
  },

  /**
   * Get the (blurry) thumbnail-sized version of an album's cover.
   */
  getThumbnail: async (album: Album): Promise<string | null> => {
    if (album.thumbnail === undefined) {
      const { thumbnailUrl } = await http.get<{ thumbnailUrl: string }>(`album/${album.id}/thumbnail`)

      album.thumbnail = thumbnailUrl
    }

    return album.thumbnail
  }
}
