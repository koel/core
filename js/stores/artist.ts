import Vue from 'vue'
import { difference, take, orderBy } from 'lodash'

import { http } from '@/services'
import stub from '@/stubs/artist'
import { use } from '@/utils'

const UNKNOWN_ARTIST_ID = 1
const VARIOUS_ARTISTS_ID = 2

export const artistStore = {
  stub,
  cache: {} as { [key: number]: Artist },

  state: {
    artists: [] as Artist[]
  },

  init (artists: Artist[]): void {
    this.all = artists

    // Traverse through artists array to get the cover and number of songs for each.
    this.all.forEach(artist => this.setupArtist(artist))
  },

  setupArtist (artist: Artist): void {
    Vue.set(artist, 'playCount', 0)
    Vue.set(artist, 'info', null)
    Vue.set(artist, 'albums', [])
    Vue.set(artist, 'songs', [])

    this.cache[artist.id] = artist
  },

  get all (): Artist[] {
    return this.state.artists
  },

  set all (value: Artist[]) {
    this.state.artists = value
  },

  byId (id: number): Artist | undefined {
    return this.cache[id]
  },

  byIds (ids: number[]): Artist[] {
    const artists = [] as Artist[]
    ids.forEach(id => use(this.byId(id), artist => artists.push(artist!)))
    return artists
  },

  add (artists: Artist | Artist[]) {
    (<Artist[]>[]).concat(artists).forEach(artist => {
      this.setupArtist(artist)
      artist.playCount = artist.songs.reduce((count, song) => count + song.playCount, 0)
      this.all.push(artist)
    })
  },

  purify (): void {
    this.compact()
  },

  /**
   * Remove empty artists from the store.
   */
  compact (): void {
    const emptyArtists = this.all.filter(artist => artist.songs.length === 0)

    if (!emptyArtists.length) {
      return
    }

    this.all = difference(this.all, emptyArtists)
    emptyArtists.forEach(artist => delete this.cache[artist.id])
  },

  isVariousArtists: (artist: Artist) => artist.id === VARIOUS_ARTISTS_ID,

  isUnknownArtist: (artist: Artist) => artist.id === UNKNOWN_ARTIST_ID,

  getSongsByArtist: (artist: Artist) => artist.songs,

  getMostPlayed (n: number = 6): Artist[] {
    // Only non-unknown artists with actual play count are applicable.
    // Also, "Various Artists" doesn't count.
    const applicable = this.all.filter(artist => {
      return artist.playCount &&
        !this.isUnknownArtist(artist) &&
        !this.isVariousArtists(artist)
    })

    return take(orderBy(applicable, 'playCount', 'desc'), n)
  },

  /**
   * Upload an image for an artist.
   *
   * @param {Artist} artist The artist object
   * @param {string} image The content data string of the image
   */
  uploadImage: async (artist: Artist, image: string): Promise<string> => {
    const { imageUrl } = await http.put<{ imageUrl: string }>(`artist/${artist.id}/image`, { image })
    artist.image = imageUrl
    return artist.image
  }
}
