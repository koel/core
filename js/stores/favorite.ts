import { difference, union } from 'lodash'
import NProgress from 'nprogress'

import { http } from '@/services'
import { alerts, pluralize } from '@/utils'

interface FavoriteStore {
  state: {
    songs: Song[]
    length: number
    fmtLength: string
  }
  all: Song[],

  toggleOne(song: Song): Promise<Song>
  add(songs: Song | Song[]): void
  remove(songs: Song | Song[]): void
  clear(): void
  like(songs: Song[]): Promise<Song[]>
  unlike(songs: Song[]): Promise<Song[]>
}

export const favoriteStore: FavoriteStore = {
  state: {
    songs: [],
    length: 0,
    fmtLength: ''
  },

  get all () {
    return this.state.songs
  },

  set all (value) {
    this.state.songs = value
  },

  toggleOne (song: Song): Promise<Song> {
    // Don't wait for the HTTP response to update the status, just toggle right away.
    // This may cause a minor problem if the request fails somehow, but do we care?
    song.liked = !song.liked
    song.liked ? this.add(song) : this.remove(song)

    return new Promise((resolve, reject) => {
      http.post('interaction/like', { song: song.id }, ({ data } : { data: Song }) => {
        // We don't really need to notify just for one song.
        resolve(data)
      }, (error: any) => reject(error))
    })
  },

  /**
   * Add a song/songs into the store.
   */
  add (songs: Song | Song[]): void {
    this.all = union(this.all, (<Song[]>[]).concat(songs))
  },

  /**
   * Remove a song/songs from the store.
   */
  remove (songs: Song | Song[]): void {
    this.all = difference(this.all, (<Song[]>[]).concat(songs))
  },

  clear (): void {
    this.all = []
  },

  like (songs: Song[]): Promise<Song[]> {
    // Don't wait for the HTTP response to update the status, just set them to Liked right away.
    // This may cause a minor problem if the request fails somehow, but do we care?
    songs.forEach(song => { song.liked = true })
    this.add(songs)

    NProgress.start()

    return new Promise((resolve, reject): void => {
      http.post('interaction/batch/like', {
        songs: songs.map(song => song.id)
      }, ({ data } : { data: Song[] }): void => {
        alerts.success(`Added ${pluralize(songs.length, 'song')} into Favorites.`)
        resolve(data)
      }, (error: any) => reject(error))
    })
  },

  unlike (songs: Song[]): Promise<Song[]> {
    songs.forEach(song => { song.liked = false })
    this.remove(songs)

    NProgress.start()

    return new Promise((resolve, reject): void => {
      http.post('interaction/batch/unlike', {
        songs: songs.map(song => song.id)
      }, ({ data } : { data: Song[] }): void => {
        alerts.success(`Removed ${pluralize(songs.length, 'song')} from Favorites.`)
        resolve(data)
      }, (error: any) => reject(error))
    })
  }
}
