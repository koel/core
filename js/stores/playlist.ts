import { difference, union } from 'lodash'
import NProgress from 'nprogress'

import stub from '@/stubs/playlist'
import { http } from '@/services'
import { alerts, pluralize } from '@/utils'
import { songStore } from '.'

export const playlistStore = {
  stub,

  state: {
    playlists: []
  },

  init (playlists) {
    this.all = playlists
  },

  get all () {
    return this.state.playlists
  },

  set all (value) {
    this.state.playlists = value
  },

  fetchSongs: playlist => {
    NProgress.start()

    return new Promise((resolve, reject) => {
      http.get(`playlist/${playlist.id}/songs`, ({ data }) => {
        playlist.songs = songStore.byIds(data)
        playlist.populated = true
        resolve(playlist)
      }, error => reject(error))
    })
  },

  byId (id) {
    return this.all.find(song => song.id === id)
  },

  /**
   * Populate the playlist content by "objectifying" all songs in the playlist.
   * (Initially, a playlist only contain the song IDs).
   *
   * @param  {Object} playlist
   */
  populateContent: playlist => (playlist.songs = songStore.byIds(playlist.songs)),

  getSongs: playlist => playlist.songs,

  /**
   * Add a playlist/playlists into the store.
   *
   * @param {Array.<Object>|Object} playlists
   */
  add (playlists) {
    this.all = union(this.all, [].concat(playlists))
  },

  /**
   * Remove a playlist/playlists from the store.
   *
   * @param  {Array.<Object>|Object} playlists
   */
  remove (playlists) {
    this.all = difference(this.all, [].concat(playlists))
  },

  store (name, songs = [], rules = null) {
    if (songs.length) {
      // Extract the IDs from the song objects.
      songs = songs.map(song => song.id)
    }

    NProgress.start()

    return new Promise((resolve, reject) => {
      http.post('playlist', { name, songs, rules }, ({ data: playlist }) => {
        playlist.songs = songs
        this.populateContent(playlist)
        this.add(playlist)
        alerts.success(`Created playlist &quot;${playlist.name}&quot;.`)
        resolve(playlist)
      }, error => reject(error))
    })
  },

  delete (playlist) {
    NProgress.start()

    return new Promise((resolve, reject) => {
      http.delete(`playlist/${playlist.id}`, {}, ({ data }) => {
        this.remove(playlist)
        resolve(data)
      }, error => reject(error))
    })
  },

  async addSongs (playlist, songs) {
    if (!playlist.populated) {
      await this.fetchSongs(playlist)
    }

    return new Promise((resolve, reject) => {
      const count = playlist.songs.length
      playlist.songs = union(playlist.songs, songs)

      if (count === playlist.songs.length) {
        resolve(playlist)
        return
      }

      NProgress.start()

      http.put(`playlist/${playlist.id}/sync`, { songs: playlist.songs.map(song => song.id) }, () => {
        alerts.success(`Added ${pluralize(songs.length, 'song')} into &quot;${playlist.name}&quot;.`)

        // Playlist's songs are not reactive right away for some reason.
        // This is a dirty hack to force reactivity.
        playlist.name = `${playlist.name} `
        playlist.name = playlist.name.trim()

        resolve(playlist)
      }, error => reject(error))
    })
  },

  removeSongs: (playlist, songs) => {
    if (playlist.is_smart) {
      return
    }

    NProgress.start()

    playlist.songs = difference(playlist.songs, songs)

    return new Promise((resolve, reject) => {
      http.put(`playlist/${playlist.id}/sync`, { songs: playlist.songs.map(song => song.id) }, () => {
        alerts.success(`Removed ${pluralize(songs.length, 'song')} from &quot;${playlist.name}&quot;.`)
        resolve(playlist)
      }, error => reject(error))
    })
  },

  update: playlist => {
    NProgress.start()

    return new Promise((resolve, reject) => {
      http.put(
        `playlist/${playlist.id}`,
        { name: playlist.name, rules: playlist.rules },
        () => {
          alerts.success(`Updated playlist &quot;${playlist.name}&quot;.`)
          resolve(playlist)
        },
        error => reject(error)
      )
    })
  }
}
