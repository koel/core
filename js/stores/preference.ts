import { userStore } from '.'
import { ls } from '@/services'

interface PreferenceStore {
  storeKey: string
  state: { [key: string]: any }
  [preferenceKey: string]: any

  init(user?: User | null): void
  setupProxy(): void
  get(key: string): any
  set(key: string, value: any): void
  save(): void
}

export const preferenceStore: PreferenceStore = {
  storeKey: '',

  state: {
    volume: 7,
    notify: true,
    repeatMode: 'NO_REPEAT',
    showExtraPanel: true,
    confirmClosing: false,
    equalizer: {
      preamp: 0,
      gains: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    },
    artistsViewMode: null,
    albumsViewMode: null,
    selectedPreset: -1,
    transcodeOnMobile: false,
    supportBarNoBugging: false,
    showAlbumArtOverlay: true
  },

  init (user?: User): void {
    const initUser = user || userStore.current
    this.storeKey = `preferences_${initUser.id}`
    this.state = Object.assign(this.state, ls.get(this.storeKey, this.state))
    this.setupProxy()
  },

  /**
   * Proxy the state properties, so that each can be directly accessed using the key.
   */
  setupProxy (): void {
    Object.keys(this.state).forEach((key: string): void => {
      Object.defineProperty(this, key, {
        get: (): any => this.get(key),
        set: (value: any): void => this.set(key, value),
        configurable: true
      })
    })
  },

  set (key: string, val: any): void {
    this.state[key] = val
    this.save()
  },

  get (key: string): any {
    return key in this.state ? this.state[key] : null
  },

  save (): void {
    ls.set(this.storeKey, this.state)
  }
}
