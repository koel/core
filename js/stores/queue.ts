import { union, difference, shuffle } from 'lodash'

export const queueStore = {
  state: {
    songs: [],
    current: null
  },

  init () {
    // We don't have anything to do here yet.
    // How about another song then?
    //
    // LITTLE WING
    // -- Jimi Hendrix
    //
    // Well she's walking
    // Through the clouds
    // With a circus mind
    // That's running wild
    // Butterflies and zebras and moonbeams and fairytales
    // That's all she ever thinks about
    // Riding with the wind
    //
    // When I'm sad
    // She comes to me
    // With a thousand smiles
    // She gives to me free
    // It's alright she said
    // It's alright
    // Take anything you want from me
    // Anything...
  },

  get all () {
    return this.state.songs
  },

  set all (songs) {
    this.state.songs = songs
  },

  /**
   * @return {?Object}
   */
  get first () {
    return this.all[0]
  },

  /**
   * @return {?Object}
   */
  get last () {
    return this.all[this.all.length - 1]
  },

  contains (song) {
    return this.all.includes(song)
  },

  /**
   * Add a list of songs to the end of the current queue.
   * @param {Object|Array.<Object>} songs The song, or an array of songs
   */
  queue (songs) {
    this.unqueue(songs)
    this.all = union(this.all, [].concat(songs))
  },

  /**
   * Add a list of songs to the top of the current queue.
   * @param {Object|Array.<Object>} songs The song, or an array of songs
   */
  queueToTop (songs) {
    this.all = union([].concat(songs), this.all)
  },

  /**
   * Replace the current queue with a list of songs.
   * @param {Object|Array.<Object>} songs The song, or an array of songs
   */
  replaceQueueWith (songs) {
    this.all = [].concat(songs)
  },

  /**
   * Queue songs right after the currently played song.
   * @param  {Array.<Object>|Object} songs
   */
  queueAfterCurrent (songs) {
    songs = [].concat(songs)

    if (!this.current || !this.all.length) {
      return this.queue(songs)
    }

    // First we unqueue the songs to make sure there are no duplicates.
    this.unqueue(songs)

    const head = this.all.splice(0, this.indexOf(this.current) + 1)
    this.all = head.concat(songs, this.all)
  },

  /**
   * @param  {Object|String|Array.<Object>} songs The song(s) to unqueue
   */
  unqueue (songs) {
    this.all = difference(this.all, [].concat(songs))
  },

  /**
   * Move some songs to after a target.
   *
   * @param  {Array.<Object>} songs  Songs to move
   * @param  {Object}     target The target song object
   */
  move (songs, target) {
    const targetIndex = this.indexOf(target)

    songs.forEach(song => {
      this.all.splice(this.indexOf(song), 1)
      this.all.splice(targetIndex, 0, song)
    })
  },

  clear () {
    this.all = []
  },

  indexOf (song) {
    return this.all.indexOf(song)
  },

  /**
   * @return {?Object}
   */
  get next () {
    if (!this.current) {
      return this.first
    }

    const index = this.all.map(song => song.id).indexOf(this.current.id) + 1

    return index >= this.all.length ? null : this.all[index]
  },

  /**
   * @return {?Object}
   */
  get previous () {
    if (!this.current) {
      return this.last
    }

    const index = this.all.map(song => song.id).indexOf(this.current.id) - 1

    return index < 0 ? null : this.all[index]
  },

  get current () {
    return this.state.current
  },

  set current (song) {
    this.state.current = song
    return this.state.current
  },

  shuffle () {
    this.all = shuffle(this.all)
    return this.all
  }
}
