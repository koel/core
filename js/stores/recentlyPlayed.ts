import NProgress from 'nprogress'
import { songStore } from '.'
import { http } from '../services'
import { remove } from 'lodash'

const EXCERPT_COUNT = 7

export const recentlyPlayedStore = {
  excerptState: {
    songs: []
  },

  state: {
    songs: []
  },

  fetched: false,

  initExcerpt (songIds) {
    this.excerptState.songs = songStore.byIds(songIds)
  },

  fetchAll () {
    NProgress.start()

    return new Promise((resolve, reject) => {
      if (this.fetched) {
        resolve(this.state.songs)
        return
      }

      http.get(`interaction/recently-played`, ({ data }) => {
        this.state.songs = songStore.byIds(data)
        resolve(this.state.songs)
      }, error => reject(error))
    })
  },

  add (song) {
    [this.state, this.excerptState].forEach(state => {
      // make sure there's no duplicate
      remove(state.songs, s => s.id === song.id)
      state.songs.unshift(song)
    })

    this.excerptState.songs.splice(EXCERPT_COUNT)
  }
}
