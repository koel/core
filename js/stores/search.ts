import { http } from '@/services'
import { songStore } from '@/stores/song'
import { albumStore } from '@/stores/album'
import { artistStore } from '@/stores/artist'

interface ExcerptSearchResult {
  songs: Array<string>
  albums: Array<number>
  artists: Array<number>
}

interface SongSearchResult {
  songs: Array<string>
}

export const searchStore = {
  state: {
    excerpt: {
      songs: [] as Song[],
      albums: [] as Album[],
      artists: [] as Artist[]
    },
    songs: [] as Song[],
  },

  excerptSearch (q: string) {
    http.get<{ [key: string]: ExcerptSearchResult }>(`search?q=${q}`).then(({ results }) => {
      this.state.excerpt.songs = songStore.byIds(results.songs)
      this.state.excerpt.albums = albumStore.byIds(results.albums)
      this.state.excerpt.artists = artistStore.byIds(results.artists)
    })
  },

  songSearch (q: string) {
    http.get<SongSearchResult>(`search/songs?q=${q}`).then(({ songs }) => {
      this.state.songs = this.state.songs.concat(songStore.byIds(songs))
    })
  },

  resetSongResultState () {
    this.state.songs = []
  }
}
