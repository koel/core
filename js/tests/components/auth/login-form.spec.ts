import Component from '@/components/auth/login-form.vue'
import { userStore } from '@/stores'
import { mock } from '@/tests/__helpers__'
import { shallow } from '@/tests/adapter'

describe('components/auth/login-form', () => {
  afterEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
  })

  it('renders properly', () => {
    expect(shallow(Component)).toMatchSnapshot()
  })

  it('triggers login when form is submitted', () => {
    const loginStub = mock(userStore, 'login')
    shallow(Component, {
      data: () => ({
        email: 'john@doe.com',
        password: 'secret'
      })
    }).submit('form')
    expect(loginStub).toHaveBeenCalledWith('john@doe.com', 'secret')
  })
})
