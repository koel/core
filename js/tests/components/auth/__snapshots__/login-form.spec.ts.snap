// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/auth/login-form renders properly 1`] = `
<form data-cy="loginForm" class="">
  <div class="logo"><img src="@/../img/logo.svg" width="156" height="auto"></div>
  <!----> <input type="email" placeholder="Email Address" autofocus="autofocus" required="required"> <input type="password" placeholder="Password" required="required">
  <btn-stub type="submit">Log In</btn-stub>
</form>
`;
